File: problems/503.下一个更大元素-ii.cpp
/*
 * @lc app=leetcode.cn id=503 lang=cpp
 *
 * [503] 下一个更大元素 II
 */

// @lc code=start
#include <iostream>
using namespace std;
#include <stack>
#include <vector>

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();

        // 现在数组是循环的，因此每个元素理论上来说都可以绕一圈来寻找，所以为了模拟这个过程，可以把数组扩大2倍，当然不是真的扩大，只是用 % 来模拟了这个过程
        stack<int>
            s;
        vector<int> res;

        // res.resize(2 * n);
        // for (int i = 2 * n - 1; i >= 0; --i) {
        //     // 比较，弹出
        //     while (!s.empty() && nums[i % n] >= s.top())
        //         s.pop();

        //     res[i] = s.empty() ? -1 : s.top();

        //     s.push(nums[i % n]);
        // }

        // res.resize(n);
        // return res;

        // 我们做一个优化，这里的res其实只用开n就可以了，因为我去模，前面的数据就会被真正需要的答案给覆盖掉
        res.resize(n);

        for (int i = 2 * n - 1; i >= 0; --i) {
            // 比较，弹出
            while (false == s.empty() && nums[i % n] >= s.top())
                s.pop();

            res[i % n] = s.empty() ? -1 : s.top();

            s.push(nums[i % n]);
        }

        return res;
    }
};
// @lc code=end

File: leetcode/python-template/leetcode/editor/cn/503.下一个更大元素-ii.py
#
# @lc app=leetcode.cn id=503 lang=python3
#
# [503] 下一个更大元素 II
#

# @lc code=start
class Solution:
    def nextGreaterElements(self, nums: list[int]) -> list[int]:
        # 让原数组拼到后面成为两倍即可解决问题
        # [1, 2, 1, 1, 2, 1]
        nextNums = [-1]*len(nums)
        s = []
        i = 2*len(nums)-1
        while i >= 0:
            while 0 != len(s) and nums[i % len(nums)] >= s[len(s)-1]:
                s.pop(len(s)-1)
            if 0 == len(s):
                nextNums[i % len(nums)] = -1
            else:
                nextNums[i % len(nums)] = s[len(s)-1]
            s.insert(len(s), nums[i % len(nums)])
            i -= 1
        return nextNums

# @lc code=end
